circuits/bigint.circom

```diff
@@ -0,0 +1,566 @@
+pragma circom 2.0.2;
+
+include "../node_modules/circomlib/circuits/comparators.circom";
+include "../node_modules/circomlib/circuits/bitify.circom";
+include "../node_modules/circomlib/circuits/gates.circom";
+
+include "bigint_func.circom";
+
+// addition mod 2**n with carry bit
+template ModSum(n) {
+    assert(n <= 252);
+    signal input a;
+    signal input b;
+    signal output sum;
+    signal output carry;
+
+    component n2b = Num2Bits(n + 1);
+    n2b.in <== a + b;
+    carry <== n2b.out[n];
+    sum <== a + b - carry * (1 << n);
+}
+
+// a - b
+template ModSub(n) {
+    assert(n <= 252);
+    signal input a;
+    signal input b;
+    signal output out;
+    signal output borrow;
+    component lt = LessThan(n);
+    lt.in[0] <== a;
+    lt.in[1] <== b;
+    borrow <== lt.out;
+    out <== borrow * (1 << n) + a - b;
+}
+
+// a - b - c
+// assume a - b - c + 2**n >= 0
+template ModSubThree(n) {
```

Comment:
`out` can be n+1 bits, if `a` is n+1 bits and `b+c` is <= n bits. Example: n=2, a=5, b=c=0 leads to out=5 and borrow=0